src/lepton/bitops.cc

```diff
@@ -159,55 +153,30 @@ abytewriter::~abytewriter( void )
 
 void abytewriter::write( unsigned char byte )
 {
-	// safety check for error
-	if ( error ) return;
-	
-	// test if pointer beyond flush threshold
-	if ( cbyte >= ( dsize - 2 ) ) {
-        if (data) {
-            unsigned char * newData = aligned_alloc(dsize *  2);
-            memcpy(newData, data, dsize);
-            dsize *= 2;
-            aligned_dealloc(data);
-            data = newData;
-        }
-		if ( data == NULL ) {
-			error = true;
-            custom_exit(ExitCode::MALLOCED_NULL);
-			return;
-		}
-	}
-	
-	// write data
-	data[ cbyte++ ] = byte;
+	write_n ( &byte, 1 );
 }
 
 /* -----------------------------------------------
 	writes n byte to abytewriter
 	----------------------------------------------- */
 
-void abytewriter::write_n( unsigned char* byte, int n )
+void abytewriter::write_n( unsigned char* src, int n )
 {
 	// safety check for error
 	if ( error ) return;
 
-	// make sure that pointer doesn't get beyond flush threshold
-	while ( ( cbyte + n ) >= ( dsize - 2 ) ) {
-        unsigned char * newData = aligned_alloc(dsize *  2);
-        memcpy(newData, data, dsize);
-        dsize *= 2;
-        aligned_dealloc(data);
-        data = newData;
-		if ( data == NULL ) {
-            error = true;
-            custom_exit(ExitCode::MALLOCED_NULL);
-			return;
+	// Extend buffer if required
+	if ( cbyte + n > dsize ) {
```

Comment:
Why was `- 2` dropped?

Reply:
The [original code](https://github.com/dropbox/lepton/blob/9c6182fcdbd604b1660fba761f373592b0d3336a/src/lepton/bitops.cc#L160):
- executed `if ( cbyte >= dsize - 2 )  extend()` before writing single byte (correct check is `cbyte+1 > dsize`)
- executed `if ( cbyte+n >= dsize - 2 )  extend()` before writing n bytes (correct check is `cbyte+n > dsize`)

Since these checks were already inconsistent, i supposed that author just overallocated a few bytes in order to simplify correctness reasoning. So I made the checks exact.

GCC sanitizer checks don't found any out-of-buffer reading situations on 50K files testset. I will also ask author whether these extra bytes are really required, but by default I suppose that it was just quick&dirty code.